lib/bedrock-item.js

```diff
@@ -0,0 +1,207 @@
+function loader (registryOrVersion) {
+  const registry =
+        typeof registryOrVersion === 'string' ? require('prismarine-registry')(registryOrVersion) : registryOrVersion
+
+  // TODO:
+  // - tests
+  // - docs
+  // - add support for older versions:
+  //   need to see how items are handled in older bedrock protocol versions,
+  //   and probably update features.json in minecraft-data with useful data
+  // - Stack ID generation
+  // - Block runtime ID field
+  // - Setting canPlaceOn and canDestroy with block/item objects?
+  // - Merge with Item class in ../index.js?
+
+  class BedrockItem {
+    /**
+         * @param {number} type
+         * @param {number} count
+         * @param {number} metadata
+         * @param {object} nbt
+         * @param {string[]} canPlaceOn
+         * @param {string[]} canDestroy
+         * @param {number} stackId
+         * @returns {BedrockItem}
+         */
+    constructor (type, count, metadata, nbt, canPlaceOn, canDestroy, stackId) {
+      if (type === null) return
+
+      this.type = type
+      this.count = count
+      this.metadata = metadata == null ? 0 : metadata
+      this.nbt = nbt || null
+
+      // TODO
+      // Only generate if on server side and not provided one
+      this.stackId = stackId ?? BedrockItem.nextStackId()
+
+      this.canPlaceOn = canPlaceOn ?? []
+      this.canDestroy = canDestroy ?? []
+
+      const itemEnum = registry.items[type]
+      if (itemEnum) {
+        this.name = itemEnum.name
+        this.displayName = itemEnum.displayName
+        if ('variations' in itemEnum) {
+          for (const variation of itemEnum.variations) {
+            if (variation.metadata === metadata) this.displayName = variation.displayName
+          }
+        }
+        this.stackSize = itemEnum.stackSize
+      } else {
+        this.name = 'unknown'
+        this.displayName = 'unknown'
+        this.stackSize = 1
+      }
+    }
+
+    /** @param {BedrockItem} item1 @param {BedrockItem} item2 */
+    static equal (item1, item2, matchStackSize = true, sameStack = false, placeable = false, destroyer = false) {
+      if (item1 == null && item2 == null) {
+        return true
+      } else if (item1 == null) {
+        return false
+      } else if (item2 == null) {
+        return false
+      } else {
+        return (
+          item1.type === item2.type &&
+                    (matchStackSize ? item1.count === item2.count : true) &&
+                    item1.metadata === item2.metadata &&
+                    JSON.stringify(item1.nbt) === JSON.stringify(item2.nbt) &&
+                    (sameStack ? item1.stackId === item2.stackId : true) &&
+                    (placeable ? item1.canPlaceOn.sort().toString() === item2.canPlaceOn.sort().toString() : true) &&
```

Comment:
too complicated, can this logic be simplified?

Reply:
Should it not be optional to check if it can be placed on / destroy the same blocks? or maybe just not check that at all?

Reply:
This can just be considered an extended part of what would normally be NBT data. You can add a `checkNbt = true` parameter and check against nbt / the bedrock `canPlaceOn` things